Track deallocated/dead in an array

tautschnig opened the pull request:
This enables tracking of all deallocations and all dead objects, and
thereby use of deallocated/dead_object predicates within assumptions.

This requires performance evaluation.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Track deallocated/dead in an array

This enables tracking of all deallocations and all dead objects, and
thereby use of deallocated/dead_object predicates within assumptions.

tautschnig commented:
![Screenshot 2022-03-29 at 20 05 15](https://user-images.githubusercontent.com/1144736/160676546-7cab5532-e21a-4fd9-9737-5fad7da603d0.png)

The above scatter plot shows current develop (y axis) vs this proposed change (x axis) on SV-COMP's MemSafety-Heap category. This one is the benchmark set with the strongest negative signal I have found so far (several others show no significant difference, some even an improvement): on this benchmark set, CBMC is now only able to solve 140 instead of 145 tasks within the resource limits. Three of those that can no longer be solved now run out of memory, two trigger a bug (fixed in #6772). This is perhaps to be expected for we end up making use of the array theory, which may generate a lot of extra constraints.

Despite this data, I'd like to see this change being considered for inclusion as it would alleviate subtle usage limitations of `__CPROVER_r_ok` (it would now be safe to be used in assumptions), and might even make it possible to get rid of `--pointer-primitive-check` at some point, which is an option that is hard to explain to users.

peterschrammel reviewed (commented):
What are the special characteristics of the benchmark that is such an outlier?

tautschnig commented:
> What are the special characteristics of the benchmark that is such an outlier?

I'd say: a lot of objects to track where constant propagation might be unable to pick out one specific object each time there's a read to the array newly introduced in this PR. Two trouble makers are https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/blob/main/c/list-ext-properties/test-0217_1.c and https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/blob/main/c/memsafety/test-0235-3.c.

There are also two tests that appear to cause an invariant failure that may point to a bug elsewhere.

tautschnig commented:
> There are also two tests that appear to cause an invariant failure that may point to a bug elsewhere.

This bug is now addressed by #6772. The more frequent use of this code via this PR just made it visible.

tautschnig committed:
Remove unnecessary __CPROVER_deallocated updates

We always allocated fresh objects, it is impossible for the condition in
these assignments to be true.
